# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/wifi/nsWifiAccessPoint.cpp
# Commit: 028af3d71e19
# Full Hash: 028af3d71e1918406a0bc3b249ef8feaa0348643
# Author: David Parks <daparks@mozilla.com>
# Date: 2023-05-13 21:42:25
# Regressor Bug: 1810421
# File Overlap Count: 2
# Description:
#   Bug 1810421: Only scan wifi when the network changes or when on mobile r=necko-reviewers,kershaw
#   
#   Simplifies the concurrent operations of wifi scanning and reduces the
#   frequency in common cases.  Wifi scanning when on mobile is reduced from
#   every 5 seconds to every minute.  Wifi scans will also happen whenever
# ==============================================================================

diff -r 35c408b9531d -r 028af3d71e19 netwerk/wifi/nsWifiAccessPoint.cpp
--- a/netwerk/wifi/nsWifiAccessPoint.cpp	Sat May 13 00:36:00 2023 +0000
+++ b/netwerk/wifi/nsWifiAccessPoint.cpp	Sat May 13 00:36:01 2023 +0000
@@ -42,40 +42,26 @@
   return NS_OK;
 }
 
-// Helper functions:
-
-bool AccessPointsEqual(nsCOMArray<nsWifiAccessPoint>& a,
-                       nsCOMArray<nsWifiAccessPoint>& b) {
-  if (a.Count() != b.Count()) {
-    LOG(("AccessPoint lists have different lengths\n"));
-    return false;
+int nsWifiAccessPoint::Compare(const nsWifiAccessPoint& o) const {
+  int ret = strcmp(mMac, o.mMac);
+  if (ret) {
+    return ret;
+  }
+  if (mSsidLen != o.mSsidLen) {
+    return (mSsidLen < o.mSsidLen) ? -1 : 1;
   }
-
-  for (int32_t i = 0; i < a.Count(); i++) {
-    LOG(("++ Looking for %s\n", a[i]->mSsid));
-    bool found = false;
-    for (int32_t j = 0; j < b.Count(); j++) {
-      LOG(("   %s->%s | %s->%s\n", a[i]->mSsid, b[j]->mSsid, a[i]->mMac,
-           b[j]->mMac));
-      if (!strcmp(a[i]->mSsid, b[j]->mSsid) &&
-          !strcmp(a[i]->mMac, b[j]->mMac) && a[i]->mSignal == b[j]->mSignal) {
-        found = true;
-      }
-    }
-    if (!found) return false;
+  ret = strncmp(mSsid, o.mSsid, mSsidLen);
+  if (ret) {
+    return ret;
   }
-  LOG(("   match!\n"));
-  return true;
+  if (mSignal == o.mSignal) {
+    return 0;
+  }
+  return (mSignal < o.mSignal) ? -1 : 1;
 }
 
-void ReplaceArray(nsCOMArray<nsWifiAccessPoint>& a,
-                  nsCOMArray<nsWifiAccessPoint>& b) {
-  a.Clear();
-
-  // better way to copy?
-  for (int32_t i = 0; i < b.Count(); i++) {
-    a.AppendObject(b[i]);
-  }
-
-  b.Clear();
+bool nsWifiAccessPoint::operator==(const nsWifiAccessPoint& o) const {
+  LOG(("nsWifiAccessPoint comparing %s->%s | %s->%s | %d -> %d\n", mSsid,
+       o.mSsid, mMac, o.mMac, mSignal, o.mSignal));
+  return Compare(o) == 0;
 }